### PR TITLE
feat(joins) Composite query execution pipeline step 2: Introduce composite query plan and dataset pipeline builder

### PR DESCRIPTION
**Pull request by fpacifici**
In order to execute a composite query we will follow the same principle of the simple query:
1) instantiate a query execution pipeline
2) this contain a query planner
3) this produces a pllan with query, processors and strategy
4) the execution pipeline executes the processing and relies on the strategy to run the query.

For the composite one.
1) the dataset instantiates an execution pipeline
2) this contains a composite planner
3) this relies on the simple planner for each entity to produce a plan for each subquery
4) These are packed into a composite plan which includes the composite query translated
5) The execution pipeline executes processing and then relies onn the composite strategy to run the composite query.

This PR introduces composite plan and dataset pipeline builder.

### CODE REVIEW

File: snuba/datasets/dataset.py [hunk 1]
```diff
@@ -47,3 +52,29 @@ def get_default_entity(self) -> Entity:
 
     def get_all_entities(self) -> Sequence[Entity]:
         return [self.get_default_entity()]
+
+    def get_query_pipeline_builder(self) -> DataSetQueryPipelineBuilder:
```
Comment (evanh) [inline snuba/datasets/dataset.py:56]:
```suggestion
    def get_query_pipeline_builder(self) -> DatasetQueryPipelineBuilder:
```

File: snuba/datasets/plans/query_plan.py [hunk 2]
```diff
@@ -2,43 +2,108 @@
...
+
+    This contains the composite query, the composite strategy and
+    the processors for all the simple subqueries.
+    SubqueryProcessors instances contain all the storage query
+    processors that still have to be executed on the subquery.
+    """
+
+    # If there is no join there would be no table alias and one
+    # single simple subquery.
+    root_processors: Optional[SubqueryProcessors]
```
Comment (evanh) [inline snuba/datasets/plans/query_plan.py:91]: When would there be a composite query with a single subquery?

Comment (fpacifici) [reply snuba/datasets/plans/query_plan.py:91]:
```
SELECT avg(soemthing) FROM (
     SELECT uniq(release) FROM sentry ..... GROUP BY platform
)
```

Comment (evanh) [reply snuba/datasets/plans/query_plan.py:93]: Oh yeah of course.

File: snuba/datasets/dataset.py [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
 from snuba.datasets.entity import Entity
+from snuba.datasets.plans.query_plan import QueryRunner
+from snuba.pipeline.query_pipeline import QueryExecutionPipeline
 from snuba.query.logical import Query
+from snuba.request import Request
 
 
-class Dataset(object):
+class Dataset:
     # TODO: s/dataset/entity/g
```
Comment (evanh) [inline snuba/datasets/dataset.py:15]: You can get rid of this too.

File: snuba/datasets/plans/query_plan.py [hunk 4]
```diff
@@ -17,32 +18,94 @@
...
+    Plan processors have to be executed only once per plan contrarily
+    to the DB Query Processors, still provided by the storage, which
     are executed by the execution strategy at every DB Query.
-    It also provides a plan execution strategy that takes care of coordinating the
-    execution of the query against the database. The execution strategy can also decide
-    to split the query into multiple chunks.
     """
 
-    query: Query
     plan_query_processors: Sequence[QueryProcessor]
```
Comment (evanh) [inline snuba/datasets/plans/query_plan.py:65]: It seems like the SubqueryProcessors class could be used here too.

**Review by evanh — APPROVED** [review]

File: snuba/datasets/plans/query_plan.py (hunk 4 above)
Comment (fpacifici) [reply snuba/datasets/plans/query_plan.py:65]: Mind if I change this on top of #1510 ? Doing it here would be quite hard to merge as these are used a lot there.

**Pull request merged**